src/codegen.cpp

```diff
@@ -3925,7 +3934,9 @@ static jl_cgval_t emit_checked_var(jl_codectx_t &ctx, Value *bp, jl_sym_t *name,
     v->setOrdering(AtomicOrdering::Unordered);
     if (tbaa)
         tbaa_decorate(tbaa, v);
-    undef_var_error_ifnot(ctx, ctx.builder.CreateIsNotNull(v), name);
+    Value *null = Constant::getNullValue(PointerType::get(ctx.builder.getContext(), 0)); // ctx.builder.CreateIntToPtr(ConstantInt::get(getSizeTy(ctx.builder.getContext()), 0), PointerType::get(ctx.builder.getContext(), 0));
+    Value *isdef = ctx.builder.CreateICmpNE(v, ctx.builder.CreateAddrSpaceCast(null, ctx.types().T_prjlvalue));
+    undef_var_error_ifnot(ctx, isdef, name);
```

Comment:
I thought this might convince LLVM that the literal pointer value really is nonnull, but that doesn't seem to be the case:
```
julia> global y::Int

julia> function g()
           global y = 1
           y
       end
g (generic function with 1 method)

julia> @code_llvm raw=true  g()
;  @ REPL[3]:1 within `g`
define i64 @julia_g_1439() #0 !dbg !5 {
top:
;  @ REPL[3]:2 within `g`
  store atomic {} addrspace(10)* addrspacecast ({}* inttoptr (i64 139877983137888 to {}*) to {} addrspace(10)*), {} addrspace(10)** inttoptr (i64 139877958293464 to {} addrspace(10)**) unordered, align 8, !dbg !7, !tbaa !8
;  @ REPL[3]:3 within `g`
  br i1 icmp ne ({} addrspace(10)* addrspacecast ({}* inttoptr (i64 139877983137888 to {}*) to {} addrspace(10)*), {} addrspace(10)* addrspacecast ({}* null to {} addrspace(10)*)), label %ok, label %err, !dbg !13

err:                                              ; preds = %top
  call void @ijl_undefined_var_error({} addrspace(12)* addrspacecast ({}* inttoptr (i64 139877984243832 to {}*) to {} addrspace(12)*)), !dbg !13
  unreachable, !dbg !13

ok:                                               ; preds = %top
  %0 = load i64, i64 addrspace(10)* addrspacecast (i64* inttoptr (i64 139877983137888 to i64*) to i64 addrspace(10)*), align 8, !dbg !13, !tbaa !14
  ret i64 %0, !dbg !13
}
```

Reply:
It is not possible to determine if the addrspacecast produces a null value from that input constant

Reply:
Ah, so the addrspace cast might not be injective? What's the proper way to tell LLVM that it can assume a null in addrspace 10 always corresponds to a literal zero?